OAuth: Invalidate GH session on failures instead of server errors

petr-muller opened the pull request:
Previously, `HandlePrStatus` failed with 500 when the user has an
invalid/stale cookie. After the fix, the behavior is the same like when
the token is successfuly extracted from the cookie but fails to
authenticate with GH. The sessions are invalidated by deleting the
cookie and `HandlePrStatus` returns indicating the user is not properly
logged in with GitHub via OAuth.

Fixes: #10063

BenTheElder commented:
/ok-to-test

petr-muller commented on prow/prstatus/prstatus.go:
```diff
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	gogithub "github.com/google/go-github/github"
+	"github.com/gorilla/sessions"
```
I don't feel good about adding an import just for the sake of the fix but I really wanted to extract session invalidation to a method (it's now used in two places) and for that I needed `sessions.Session` type. Maybe there's some better solution for that.

fejta replied:
What good is a package if it doesn't import every other package? At least this library is already vendored.

petr-muller commented:
/assign @fejta

petr-muller commented on prow/prstatus/prstatus_test.go:
```diff
@@ -119,6 +119,53 @@ func TestHandlePrStatusWithoutLogin(t *testing.T) {
+	testCases := []struct {
+		existingCookie *http.Cookie
+		expectedData   UserData
+	}{{
```
We probably don't. I have an unhealthy tendency to future-proof things. I've flattened the testcase in an amend.

petr-muller commented:
> I'm not convinced the way that we handle cookies and sessions in this code is ideal or strictly correct.

That's what I was thinking too, but I am not a web programmer good enough to try to clean it up. 
Thanks for the reviews, I addressed the feedback about the testcase in an amend.

BenTheElder commented:
/approve

Katharine commented:
/lgtm
